### PR TITLE
Fix snippet option conflicts web_editor / website / mass_mailing

### PR DESCRIPTION
**Pull request by detrouxdev**
**WILL BE BACKPORTED IN 16.0**

### CODE REVIEW

**Comment by detrouxdev** [comment]
@qsm-odoo Can I have your opinion on my approach? I don't know if adding a `showUI` property to the options is really a good idea or if I should change the signature of updateUIVisibility to return false if the option is hidden. (Or if I should do that here).

I did check the overrides of the method and found 2 things that could be improved / fixed. I've put them in separate commits.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -3529,8 +3529,8 @@ const SnippetOptionWidget = Widget.extend({
             return Promise.all(proms);
         });
 
-        const showUI = await this._computeVisibility();
-        this.el.classList.toggle('d-none', !showUI);
+        this.showUI = await this._computeVisibility();
+        this.el.classList.toggle('d-none', !this.showUI);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3533]: Nah I don't think that's a good idea. The most common method is `_computeWidgetVisibility`, which we will likely use for options which only have one widget. 

But about this whole commit purpose: did not it work at some point? I am 99% sure I did not merge the "page options" work with the section showing up all the time :grimacing:

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js [hunk 2]
```diff
@@ -147,6 +130,14 @@ options.registry.ImageTools.include({
         }
         return this._super(...arguments);
     },
+
+    _computeWidgetVisibility(widgetName, params) {
+        // Transform is _very_ badly supported in mail clients. Hide the option.
+        if (widgetName === 'image_transform_opt') {
+            return this.snippets && this.snippets.split('.')[0] !== "mass_mailing";
+        }
+        return this._super(...arguments);
```
Comment (qsm-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_snippets.js:139]: Whoa... why would this file affect the website editor? Is not the mass mailing editor still inside an iframe? Why is this fine loaded at all for the website edition?

File: addons/website/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -2047,8 +2048,12 @@ options.registry.HeaderNavbar = options.Class.extend({
         // should be done another way in another place somehow...
         const noHamburgerWidget = this.findWidget('no_hamburger_opt');
         const noHamburgerHidden = noHamburgerWidget.$el.hasClass('d-none');
-        if (noHamburgerHidden && noHamburgerWidget.isActive()) {
+        // The noHamburgerWidget has 2 methods, including one which will only
+        // have its correct value once the page is reloaded. Therefore, we need
+        // to make sure the method call is only done once.
+        if (noHamburgerHidden && noHamburgerWidget.isActive() && !this._hamburgerAutoUpdated) {
             this.findWidget('default_hamburger_opt').enable();
+            this._hamburgerAutoUpdated = true;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2056]: Your commit message seems great but I don't understand the issue anyway :grimacing: Maybe we'll talk about it at the farm...

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

**Comment by detrouxdev** [comment]
@qsm-odoo this PR is ready for review! 😄

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 4]
```diff
@@ -488,7 +488,7 @@ export class MassMailingHtmlField extends HtmlField {
                 class: 'container o_mail_wrapper o_mail_regular oe_unremovable',
             });
             $newWrapperContent = $('<div/>', {
-                class: 'col o_mail_no_options o_mail_wrapper_td bg-white oe_structure o_editable'
+                class: 'col o_mail_no_options o_mail_wrapper_td bg-white oe_structure o_editable oe_unremovable'
```
Comment (qsm-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_html_field.js:491]: What does this have to do with this commit?

File: addons/mass_mailing/static/src/js/mass_mailing_snippets.js [hunk 5]
```diff
@@ -95,36 +95,17 @@ options.registry.mass_mailing_sizing_x = options.Class.extend({
...
-options.registry.BackgroundImage = options.registry.BackgroundImage.extend({
+options.registry.BackgroundImageMassMailing = options.registry.BackgroundImage.extend({
     start: function () {
         this._super();
-        if (this.snippets && this.snippets.split('.')[0] === "mass_mailing") {
-            var $table_target = this.$target.find('table:first');
-            if ($table_target.length) {
-                this.$target = $table_target;
-            }
+        var $table_target = this.$target.find('table:first');
```
Comment (qsm-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_snippets.js:101]: If you touch a line with a `var`, use const or let instead. Also use camelCase instead of skake_case

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@detrouxdev For future PR, try creating new ones for commits which are unrelated to the rest ;) Here we have a 5 commits branch named "hide-page-options" with 10 lines over 1700 which are related to that. As I also have to split it for 16.0 and master, the discussions will be split... well you understand :wink: 

- I created a PR for 16.0 with 3 out the 5 commits here, which I will r+ now: https://github.com/odoo/odoo/pull/101638
- I am push-forcing this PR (just after sending my review), to keep only the 2 remaining commits (which will need some rework and precisions, see review comments) -> this should stay in this PR as it will need to be merged in master first and backported in 16.0 later for robodoo to agree with me. Try to finish this PR for master ASAP following my comments, the later we merge this (new views, etc), the more (rightfully) angry people we will have as this will destroy already created 16.0 instances :grimacing: 
- I added some comments on commit which I moved to the 16.0 PR, make sure to consider them for future fixes (but maybe after other fixes more urgent in our pad)

File: addons/mass_mailing/views/snippets_themes.xml [hunk 6]
```diff
@@ -290,7 +290,485 @@
...
+    <!-- Transform is _very_ badly supported in mail clients -->
+    <xpath expr="//we-button[@data-name='image_transform_opt']" position="replace" />
+
+    <xpath expr="//div[@data-js='ImageTools']" position="attributes">
+        <attribute name="data-js">ImageToolsMassMailing</attribute>
+    </xpath>
+
+    <!-- Adding new Snippet Options -->
+
+    <xpath expr="." position="inside">
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/views/snippets_themes.xml:313]: Using an external template like "extra_snippet_options" means having to change every options added by specific snippet (grep "website.snippet_options") I don't mind it but it will add a bigger diff and IMO might make it harder to understand which template to target when trying to extend/modify options from an external module.

For now I went with the solution of undoing the indent I did, what do you think?

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js (hunk 4 above)
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: Moved it here: https://github.com/odoo/odoo/pull/101700

File: addons/mass_mailing/views/snippets_themes.xml [hunk 7]
```diff
@@ -290,7 +290,485 @@
...
 
-<template id="snippet_options_extra_shapes" inherit_id="web_editor.snippet_options">
+<template id="snippet_options_background_options" inherit_id="web_editor.snippet_options_background_options" primary="True">
+    <xpath expr="//div[@data-js='BackgroundImage']" position="attributes">
+        <attribute name="data-js">BackgroundImageMassMailing</attribute>
+    </xpath>
+</template>
+<!-- Snippet themes Options -->
+<template id="snippet_options" inherit_id="web_editor.snippet_options" primary="True">
+    <!-- t-set -->
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/views/snippets_themes.xml]: Interesting: in my IDE this doesn't get flagged as a line changed so I didn't pay much attention. Clearly it is though so I adapted the comment to say `Mass Mailing snippet options` tho it's self explanatory IMO.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -4112,7 +4112,7 @@ const SnippetOptionWidget = Widget.extend({
                 this.options.wysiwyg.odooEditor.historyStep();
             }
 
-            if (previewMode) {
+            if (previewMode || requiresReload) {
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: > I'll merge this one like that to avoid the dead lock but this should really be better: if you don't reclick on the navbar after the first reload... you header is somehow in an inconsistent state? 

Clearly this flow is weird, but it was like that previously (worse even since clicking on the navbar would show a random popup with no context), so my aim was just to restore it.

It could be improved in two ways:
- Since reloading the editor can take a parameter to start on a specific snippet (see reloadSelector in the wysiwywg_adapter) we could find a way to pass the navbar selector so that it automatically clicks on it. It will still be done in 2 times (template selector, then no_hamburger selector) but it won't require user interaction
- Find a way for this to be done in backend, when a view is enabled, some CSS variables get activated/deactivated as well. If it's already possible, I'm not aware of it and it's my bad, this fix could have been better.

> following the flow you described, should not the no_hamburger_opt option be entirely hidden at the end? (that's why we had to disable it with the hack)... that's not the case, needs a fix 😬

Weird, in my tests it worked, I will investigate further.

File: addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js [hunk 9]
```diff
@@ -427,6 +427,14 @@ export class WysiwygAdapterComponent extends ComponentAdapter {
...
     _websiteRootEvent(type, eventData = {}) {
         const websiteRootInstance = this.websiteService.websiteRootInstance;
+        // If the websiteRootInstance is gone but an event still tries to reach it
+        // prevent a traceback by denying the event.
+        if (!websiteRootInstance) {
+            if (eventData.onFailure) {
+                return eventData.onFailure();
+            }
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js]: Indeed it is hiding a few bugs, just like the window.location.reload() were hiding a few before.
What I could gather was the following:
If, when selecting an option that reloads the page, your mouse ends on an option that triggers a preview, a widgets_start_request will be triggered, which is what caused the random tracebacks. Maybe there is a race condition between the reactive variable ([see here](https://github.com/odoo/odoo/blob/849856843aeab04cd6dc562ca34142a5db6b7e90/addons/website/static/src/components/editor/editor.js#L101)) properly dismounting and destroying the editor and the iframe being reloaded.

A few things mitigate this race condition, as we do not reload the iframe until a save is done, which should destroy any editor alive. Though I cannot say the same for clicking on cancel.

 I'll investigate further as well.

**Comment by detrouxdev** [comment]
@qsm-odoo This PR is ready 😄
